src/redux/reducers/searchText.js

```diff
@@ -0,0 +1,198 @@
+import { handleActions } from 'redux-actions';
+import actionTypes from '../actionTypes';
+import { searchMessages } from '../../utils/searchMessages';
+
+const defaultState = {
+  search: false,
+  searchState: {
+    markedMessagesByID: {},
+    markedIDs: [],
+    markedIndices: [],
+    scrollLimit: 0,
+    scrollIndex: -1,
+    scrollID: null,
+    caseSensitive: false,
+    searchIndex: 0,
+    searchText: '',
+  },
+};
+
+export default handleActions(
+  {
+    [actionTypes.SEARCH_TEXT_CHANGED](state, { payload }) {
+      // TODO: Call close snackbar from Top Bar
+      const { messages, messagesByID, searchText } = payload;
+      let { searchState } = state;
+      // const searchText = event.target.value;
+      if (searchText.trim()) {
+        const markingData = searchMessages(
+          messages,
+          messagesByID,
+          searchText,
+          searchState.caseSensitive,
+        );
+        const { markedMessagesByID, markedIDs, markedIndices } = markingData;
+        searchState = {
+          markedMessagesByID: markedMessagesByID,
+          markedIDs: markedIDs,
+          markedIndices: markedIndices,
+          scrollLimit: markedIDs.length,
+          scrollIndex: 0,
+          scrollID: markedIDs[0],
+          caseSensitive: searchState.caseSensitive,
+          searchIndex: 1,
+          searchText: searchText,
+        };
+        if (markedIDs.length === 0 && searchText.trim()) {
+          // if no Messages are marked(i.e no result) and the search query is not empty
+          // TODO: Open snackbar from topbar
+          // openSnackBar({
+          //   snackBarMessage: 'No result found!',
+          //   snackBarDuration: 3000,
+          // });
+        }
+      } else {
+        searchState = {
+          markedMessagesByID: messagesByID,
+          markedIDs: [],
+          markedIndices: [],
+          scrollLimit: 0,
+          scrollIndex: -1,
+          scrollID: null,
+          caseSensitive: searchState.caseSensitive,
+          searchIndex: 0,
+          searchText: '',
+        };
+      }
+      return {
+        ...state,
+        searchState,
+      };
+    },
+    [actionTypes.SEARCH_PREVIOUS_ITEM](state, { payload }) {
+      // TODO: Find a way to pass ref for ul
+      let { ul } = payload;
+      const { searchState } = state;
+      let newIndex = searchState.scrollIndex + 1;
+      let newSearchCount = searchState.searchIndex + 1;
+      let indexLimit = searchState.scrollLimit;
+      let markedIDs = searchState.markedIDs;
+      if (newSearchCount <= 0) {
```

Comment:
Is this condition required

Reply:
@akshatnitd I'll work on optimizing prev and next search in this PR, I'll try to get rid of redundancies.